torch/lib/c10d/ProcessGroupWrapper.cpp

```diff
@@ -23,46 +26,74 @@ struct CollectiveFingerPrint {
   // Ref to input tensors, if given, of the collective. If given, shapes will be
   // checked across processes to ensure valid input into the collective.
   const std::vector<at::Tensor>& input_tensors_;
+  // input tensor data types
+  std::vector<int> tensor_dtypes_;
+  // input tensor device types
+  std::vector<int8_t> tensor_device_types_;
   explicit CollectiveFingerPrint(
       OpType op_type,
       const std::vector<at::Tensor>& input_tensors)
-      : op_type_(op_type), input_tensors_(input_tensors) {}
-
-  // Verifies a given int is the same across processes.
-  void verify_num(
-      int64_t value,
-      const c10::intrusive_ptr<ProcessGroup>& pg,
-      const std::string& failureMsg) {
-    auto tensor = at::full({1}, value, at::TensorOptions().dtype(at::kLong));
-    std::vector<at::Tensor> tensors;
-    tensors.reserve(pg->getSize());
-    for (int i = 0; i < pg->getSize(); ++i) {
-      tensors.emplace_back(at::zeros_like(tensor));
+      : op_type_(op_type), input_tensors_(input_tensors) {
+    tensor_dtypes_.reserve(input_tensors.size());
+    tensor_device_types_.reserve(input_tensors.size());
+    for (const at::Tensor& t : input_tensors_) {
+      tensor_dtypes_.push_back(static_cast<int8_t>(t.dtype().toScalarType()));
+      tensor_device_types_.push_back(static_cast<int8_t>(t.device().type()));
     }
-    std::vector<std::vector<at::Tensor>> out_tensors({tensors});
-    std::vector<at::Tensor> inp_tensors({tensor});
-    pg->allgather(out_tensors, inp_tensors)->wait();
-    std::unordered_set<int64_t> gathered;
-    for (const auto& t : out_tensors[0]) {
-      auto n = t.item().to<int64_t>();
-      gathered.insert(n);
-      if (gathered.size() > 1) {
-        TORCH_CHECK(false, failureMsg);
+  }
+
+  // Logs collective information in case of a failure.
+  friend std::ostream& operator<<(
+      std::ostream& output,
+      const CollectiveFingerPrint& collective_fingerprint);
+
+  at::Tensor serialize_fingerprint() {
+    std::vector<int64_t> data;
+    // OpType
+    data.push_back(static_cast<int64_t>(op_type_));
+    // Shapes
+    for (const auto& tensor : input_tensors_) {
+      auto sizes = tensor.sizes().vec();
+      for (const auto& s : sizes) {
+        data.push_back(s);
       }
     }
+    // tensor dtypes
+    for (const auto& type : tensor_dtypes_) {
+      data.push_back(type);
+    }
+    // device types
+    for (const auto& d : tensor_device_types_) {
+      data.push_back(d);
+    }
+    // Serialize data into tensor
+    int64_t data_size = data.size();
+    at::DataPtr data_ptr((void*)std::move(data.data()), c10::Device(c10::kCPU));
```

Comment:
The reason why this does not work is because at the end of the call the `data` variable gets destructed. So the pointer you retrieve here becomes dangled. On a second thought this can be solved even better with the new fluent API I introduced a while ago (#54530) Using it you can instantiate the tensor like:

```c++
auto data = std::make_unique<std::vector<int64_t>>();
...
at::Tensor tensor = at::for_blob(data->data(), {data_size})
  .context(data.release(), [](void *ctx) { delete static_cast<std::vector<int64_t>*>(ctx); })
  .make_tensor();
```

This version avoids any cloning and directly uses the `std::vector<>` as the storage backend of the tensor.

Reply:
@cbalioglu Sounds good, I'll try this API out. Although just curious, the way this is currently written with `c10::Storage`, there's no way to make `Storage`/tensor take ownership of `data`?

Reply:
Sure, you can also use the `Storage` approach. The missing piece in your code is the `ContextDeleter` that you should pass to `DataPtr`. Right now you pass the naked data pointer and as I mentioned it gets free'd at the end of the call. You should dynamically allocate your `std::vector<>` (e.g. via `std::make_unique<>`) and pass its pointer as `context`, and its deleter (basically a function/lambda with `delete ptr`) as `context_deleter` so that `DataPtr` takes ownership of your `std::vector<>`.

However my second suggestion with the fluent API in fact does this exact same thing behind the scenes for you. So there is no need for you to instantiate a `Storage` object yourself.